lib/storage/index.js

```diff
@@ -0,0 +1,239 @@
+var assign = require('object-assign');
+var fs = require('fs-extra');
+var path = require('path');
+var mime = require('mime-types');
+var namefunctions = require('./namefunctions');
+
+/**
+	The Adapter Compatibility Level is used to ensure that provided adapters
+	have been designed to work with the version of Storage implemented in the
+	version of Keystone being used.
+
+	The Adapter Compatibility Level must be an exact match or an error will be
+	thrown when the Storage instance is initialised.
+*/
+var ADAPTER_COMPATIBILITY_LEVEL = 1;
+
+/**
+	This is the default storage schema for file fields. These paths will be
+	persisted to the database and represent the value of the file field.
+
+	Storage Adapters can specify their own schema properties but must implement
+	the required schema fields (see array below).
+*/
+var SCHEMA_TYPES = {
+	// filename: String,		// the filename, without the full path
+	size: Number,		// the size of the file
+	mimetype: String,		// the mime type of the file
+	path: String,		// the path (e.g directory) the file is stored in; not the full path to the file
+	originalname: String,	// the original (uploaded) name of the file; useful when filename generated
+	url: String,			// publicly accessible URL of the stored file
+
+
+	// bucket: String,
+};
+
+/**
+	This is the default set of schema fields that are enabled. The storage
+	options can override these (must be supported by the adapter, or the
+	adapter must use the default schema implementation)
+*/
+var SCHEMA_FIELD_DEFAULTS = {
+	// filename: true,
+	size: true,
```

Comment:
How about requiring a hash, like md5? That way, etags can be generated and verified without hitting the storage adapter.